pin.js

```diff
@@ -0,0 +1,121 @@
+import { CID } from 'multiformats'
+import http from 'node:http'
+import { getPkg, getClient } from './lib.js'
+
+/**
+ * a pinning service api on your localhost
+ *
+ * ## Example
+ *   w3 ps --port 1337
+ */
+export async function startPinService ({ port, host = '127.0.0.1', key }) {
+  const pkg = getPkg()
+  const pinCache = new Map()
+  const client = await getClient()
+  const whoami = client.agent().did()
+  const token = key ?? whoami
+  const api = http.createServer(async (req, res) => {
+    if (req.headers.authorization !== `Bearer ${token}`) {
+      return send({ res, status: 401, body: { error: { reason: 'Unauthorized; access token is missing or invalid' } } })
+    }
+    const { pathname } = new URL(req.url, `http://${req.headers.host}`)
+    if (pathname === '/' || pathname === '') {
+      return send({ res, body: { service: 'w3', version: pkg.version } })
+    }
+    if (req.method === 'POST' && pathname === '/pins') {
+      const body = await getJsonBody(req)
+      const pinStatus = await addPin({ ...body, client })
+      pinCache.set(pinStatus.requestid, pinStatus)
+      return send({ res, body: pinStatus })
+    }
+    if (req.method === 'GET' && pathname.startsWith('/pins/')) {
+      const requestid = pathname.split('/').at(2)
+      const pinStatus = pinCache.get(requestid)
+      if (pinStatus) {
+        return send({ res, body: pinStatus })
+      }
+      return send({ res, status: 404, body: { error: { reason: 'Not Found', details: requestid } } })
+    }
+    return send({ res, status: 501, body: { error: { reason: 'Not Implmented', details: `${req.method} ${pathname}` } } })
+  })
+  api.listen(port, host, () => {
+    console.log(`⁂ IPFS Pinning Service on http://127.0.0.1:1337
+
+## Add w3 as a remote
+$ ipfs pin remote service add w3 'http://${host}:${port}' '${token}'
+
+## Pin to w3
+$ ipfs pin remote add --service w3 <cid>
+
+## Waiting for requests`)
+  })
+}
+
+/**
+ * @param {object} config
+ * @param {import('@web3-storage/w3up-client').Client} confg.client
+ * @param {string} config.cid
+ * @param {string} [config.ipfsGatewayUrl]
+ * @param {AbortSignal} [config.signal]
+ */
+export async function addPin ({ client, cid, ipfsGatewayUrl = 'http://127.0.0.1:8080', signal }) {
+  const rootCID = CID.parse(cid)
+  const ipfsUrl = new URL(`/ipfs/${cid}?format=car`, ipfsGatewayUrl, { signal })
+  const res = await fetch(ipfsUrl)
+  const storedCID = await client.uploadCAR({ stream: () => res.body }, {
```

Comment:
We should tweak the w3up-client `uploadCAR` to allow users to pass a stream, as currently it requires a blob but just calls stream on it, and it means we can't pass a stream directly, and need this madness.